h5py/tests/test_file.py

```diff
@@ -840,6 +842,91 @@ def test_swmr_mode_consistency(self):
         fid.close()
 
 
+@pytest.mark.skipif(
+    h5py.version.hdf5_version_tuple < (1, 12, 1) and (
+    h5py.version.hdf5_version_tuple[:2] != (1, 10) or h5py.version.hdf5_version_tuple[2] < 7),
+    reason="Requires HDF5 >= 1.12.1 or 1.10.x >= 1.10.7")
+@pytest.mark.skipif("HDF5_USE_FILE_LOCKING" in os.environ,
+                    reason="HDF5_USE_FILE_LOCKING env. var. is set")
+class TestFileLocking:
+    """Test h5py.File file locking option"""
+
+    def test_reopen(self, tmp_path):
+        """Test file locking when opening twice the same file"""
+        fname = tmp_path / "test.h5"
+
+        with h5py.File(fname, mode="w", locking=True) as f:
+            f.flush()
+
+            # Opening same file in same process without locking is expected to fail
+            with pytest.raises(OSError):
+                with h5py.File(fname, mode="r", locking=False) as h5f_read:
+                    pass
+
+            with h5py.File(fname, mode="r", locking=True) as h5f_read:
+                pass
+
+            with h5py.File(fname, mode="r", locking='best-effort') as h5f_read:
+                pass
+
+    def test_unsupported_locking(self, tmp_path):
+        """Test with erroneous file locking value"""
+        fname = tmp_path / "test.h5"
+        with pytest.raises(ValueError):
+            with h5py.File(fname, mode="r", locking='unsupported-value') as h5f_read:
+                pass
+
+    def test_unsupported_hdf5_version(self, tmp_path):
+        """Test simulating an unsupported version of HDF5"""
+        hdf5_version = h5py._hl.files.hdf5_version
+        h5py._hl.files.hdf5_version = 1, 10, 6
+
+        try:
+            fname = tmp_path / "test.h5"
+            with pytest.raises(ValueError):
+                with h5py.File(fname, mode="r", locking=True) as h5f_read:
+                    pass
+
+            with pytest.raises(ValueError):
+                h5py._hl.files.make_fapl(None, None, None, None, None, locking=True)
+
+        finally:
+            h5py._hl.files.hdf5_version = hdf5_version
+
+
+    def test_multiprocess(self, tmp_path):
+        """Test file locking option from different concurrent processes"""
+        fname = tmp_path / "test.h5"
+
+        def open_in_subprocess(filename, mode, locking):
+            """Try to open HDF5 file a subprocess and return CompletedProcess"""
+            escaped_filename = "\\\\".join(str(filename).split("\\"))
+            h5py_import_dir = pathlib.Path(h5py.__file__).parent.parent
+            escaped_h5py_import_dir = "\\\\".join(str(h5py_import_dir).split("\\"))
+
+            process = subprocess.run(
+                [
+                    sys.executable,
+                    "-c",
+                    f"""
+import sys
+sys.path.insert(0, '{escaped_h5py_import_dir}')
+import h5py
+f = h5py.File('{escaped_filename}', mode='{mode}', locking={locking})
```

Comment:
```suggestion
sys.path.insert(0, {h5py_import_dir!r})
import h5py
f = h5py.File({filename!r}, mode={mode!r}, locking={locking})
```

`!r` is formatting shorthand for `repr()`, which will do the necessary escaping & quoting to make a string literal, which is what we want here:

```
>>> a = "C:\\foo"
>>> f"{a}"
'C:\\foo'
>>> f"{a!r}"
"'C:\\\\foo'"
```